daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala

```diff
@@ -48,6 +51,72 @@ object XMLTextInfosetInputter {
     fact.setProperty(XMLInputFactory.IS_VALIDATING, false) // no DTD validation
     fact
   }
+
+  lazy val xmlOutputFactory = {
+    val fact = new com.ctc.wstx.stax.WstxOutputFactory()
+    fact
+  }
+
+  /**
+   * When the stringAsXml runtime property is set to true we treat the simple
+   * content as if it were XML and output it to the infoset. We use this value
+   * as the runtime property name. Additionally, when we output the XML we need
+   * a wrapper element to reset the default XML namespace in case the infoset
+   * has a default namespace. We also use this variable as the name of that
+   * wrapper element.
+   */
+  lazy val stringAsXml = "stringAsXml"
+
+  /**
+   * Read an event from an XMLStreamReader and write the same event to an
+   * XMLStreamWriter. For example, this can be used for convert XML from one
+   * type of input to a different type (e.g. InputStream to String), check for
+   * well-formed XML, and perform some basic normalization. Example usage of
+   * this function is to call this in a loop like this:
+   *
+   *   while (xsr.hasNext()) {
+   *     XMLTextInfoset.writeStreamEvent(xsr, xsw)
+   *     xsr.next()
+   *   }
+   *
+   * This approach allows callers to skip certain events by conditionally
+   * calling this function.
+   */
+  def writeXMLStreamEvent(xsr: XMLStreamReader, xsw: XMLStreamWriter): Unit = {
+    xsr.getEventType() match {
+      case START_ELEMENT => {
+        xsw.writeStartElement(
+          xsr.getPrefix(),
+          xsr.getLocalName(),
+          xsr.getNamespaceURI())
+        for (i <- 0 until xsr.getNamespaceCount()) {
+          xsw.writeNamespace(
+            xsr.getNamespacePrefix(i),
+            xsr.getNamespaceURI(i))
+        }
+        for (i <- 0 until xsr.getAttributeCount()) {
+          xsw.writeAttribute(
+            xsr.getAttributePrefix(i),
+            xsr.getAttributeNamespace(i),
+            xsr.getAttributeLocalName(i),
+            xsr.getAttributeValue(i))
+        }
+      }
+      case END_ELEMENT => xsw.writeEndElement()
+      case CHARACTERS => xsw.writeCharacters(xsr.getText())
+      case COMMENT => xsw.writeComment(xsr.getText())
+      case CDATA => xsw.writeCData(xsr.getText())
+      case PROCESSING_INSTRUCTION => xsw.writeProcessingInstruction(xsr.getPITarget(), xsr.getPIData())
+      case ENTITY_REFERENCE => xsw.writeEntityRef(xsr.getLocalName())
+      case DTD => xsw.writeDTD(xsr.getText())
```

Comment:
Definitely need a comment (and documentation needs to mention this also) that DTDs are allowed and preserved in the embedded payload XML strings.

Reply:
So that's maybe worth considering if we want to support that. I assumed so, but the logic in the infosetter and outputter make it pretty straightforward if there's anything that could cause issues.

For example, if we allow DTD's that define new entities, is it possible that could cause issues in other parts of the infoset if they then use those entities? Before this change, one could very safely use our infosets because they were very limited (only complex types, and simple types, and everything escape correctly). But with this change, infosets are much less safe, one could maybe insert a XML Bomb in the infoset.

Reply:
Hmmm. giving this a bit more thought. A DTD has to be at the start of the XML right? So it doesn't make sense I think now, to preserve a DTD in an XML thunk that is going to be inside another XML document as a child element somewhere. 

I think we just strip out DTDs, and document that we do.

Reply:
Yep, makes sense, I'll make that change.